File: worker/src/worker.ts
// Entry point. Rust spawns this process, writes JSON lines to stdin and reads
// JSON lines from stdout. Anything else (SDK logs, diagnostics) goes to stderr.
import { createInterface } from 'node:readline';
import { readFileSync } from 'node:fs';
import { configureLogging } from '@strands-agents/sdk';
import { MAX_LINE_BYTES, PROTOCOL_VERSION, encodeOutbound, parseInbound, type Outbound, type RunRequest } from './protocol.ts';
import { createModel, modelSettingsFromEnv } from './model.ts';
import { executeRun, type ToolOutcome } from './run.ts';

// stdout is the protocol channel. Nothing may print there except encoded messages.
const stdoutWrite = process.stdout.write.bind(process.stdout);
console.log = (...args: unknown[]) => console.error(...args);
console.info = (...args: unknown[]) => console.error(...args);
configureLogging({ debug() {}, info() {}, warn: (...args) => console.error('[strands]', ...args), error: (...args) => console.error('[strands]', ...args) });

const emit = (message: Outbound) => { stdoutWrite(encodeOutbound(message)); };

type Pending = { resolve: (outcome: ToolOutcome) => void; cleanup: () => void };
type Current = { runId: string; abort: AbortController; pending: Map<string, Pending>; finished: Promise<void> };
let current: Current | null = null;
let shuttingDown = false;

function sdkVersion(): string {
  try {
    const manifest = new URL('../node_modules/@strands-agents/sdk/package.json', import.meta.url);
    return String((JSON.parse(readFileSync(manifest, 'utf8')) as { version: string }).version);
  } catch { return 'unknown'; }
}

function startRun(request: RunRequest) {
  if (current) {
    emit({ type: 'failed', run_id: request.run_id, code: 'worker_busy', error: 'Worker already has an active run' });
    return;
  }
  let settings;
  try { settings = modelSettingsFromEnv(); } catch (error) {
    emit({ type: 'failed', run_id: request.run_id, code: 'model_not_configured', error: error instanceof Error ? error.message : 'Model is not configured' });
    return;
  }
  const abort = new AbortController();
  const pending = new Map<string, Pending>();
  const run: Current = { runId: request.run_id, abort, pending, finished: Promise.resolve() };
  current = run;
  const io = {
    emit,
    requestTool(callId: string, name: string, input: unknown, signal: AbortSignal) {
      return new Promise<ToolOutcome>((resolve) => {
        const onAbort = () => finish({ error: 'Tool request cancelled' });
        const finish = (outcome: ToolOutcome) => { pending.delete(callId); signal.removeEventListener('abort', onAbort); resolve(outcome); };
        if (signal.aborted) { finish({ error: 'Tool request cancelled' }); return; }
        pending.set(callId, { resolve: finish, cleanup: () => signal.removeEventListener('abort', onAbort) });
        signal.addEventListener('abort', onAbort, { once: true });
        emit({ type: 'tool_call', run_id: request.run_id, call_id: callId, name, input });
      });
    },
  };
  run.finished = executeRun(request, createModel(settings, request.limits.output_tokens), io, abort.signal)
    .catch((error: unknown) => { emit({ type: 'failed', run_id: request.run_id, code: 'worker_error', error: error instanceof Error ? error.message.slice(0, 500) : 'Worker error' }); })
    .finally(() => { for (const entry of pending.values()) entry.cleanup(); pending.clear(); if (current === run) current = null; if (shuttingDown) exit(0); });
}

function exit(code: number) {
  process.exitCode = code;
  setTimeout(() => process.exit(code), 50).unref();
}

async function shutdown() {
  if (shuttingDown) return;
  shuttingDown = true;
  if (current) { current.abort.abort(); await Promise.race([current.finished, new Promise((resolve) => setTimeout(resolve, 3000))]); }
  exit(0);
}

const input = createInterface({ input: process.stdin, crlfDelay: Infinity });
input.on('line', (line) => {
  if (Buffer.byteLength(line) > MAX_LINE_BYTES) { console.error('[worker] dropped oversized line'); return; }
  if (!line.trim()) return;
  let message;
  try { message = parseInbound(line); } catch (error) { console.error(`[worker] invalid message: ${error instanceof Error ? error.message : error}`); return; }
  switch (message.type) {
    case 'run': startRun(message); break;
    case 'tool_result': {
      const entry = current && current.runId === message.run_id ? current.pending.get(message.call_id) : undefined;
      if (!entry) { console.error('[worker] ignored tool_result for unknown call'); break; }
      entry.resolve(message.error !== undefined ? { error: message.error } : { result: message.result });
      break;
    }
    case 'cancel': if (current && current.runId === message.run_id) current.abort.abort(); break;
    case 'shutdown': void shutdown(); break;
  }
});
input.on('close', () => { void shutdown(); });
process.on('SIGTERM', () => { void shutdown(); });
process.on('SIGINT', () => { void shutdown(); });

emit({ type: 'ready', protocol: PROTOCOL_VERSION, sdk: sdkVersion() });

File: src/worker.rs
//! Supervises the Strands worker child process and speaks its line-delimited
//! JSON protocol. Rust keeps database credentials, workspace authority and
//! every tool implementation; the worker only orchestrates the model.
use crate::{
    config::Config,
    error::{Error, Result},
};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::{process::Stdio, time::Duration};
use tokio::{
    io::{AsyncBufReadExt, AsyncWriteExt, BufReader},
    process::{Child, ChildStdin, Command},
    sync::mpsc,
    task::JoinHandle,
};

pub const PROTOCOL_VERSION: u64 = 1;
const MAX_LINE_BYTES: usize = 1_048_576;
const READY_TIMEOUT: Duration = Duration::from_secs(30);
const SHUTDOWN_GRACE: Duration = Duration::from_secs(3);

#[derive(Debug, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Inbound<'a> {
    Run {
        run_id: &'a str,
        kind: &'a str,
        system_prompt: &'a str,
        message: &'a str,
        history: Vec<HistoryTurn>,
        tools: &'a [Value],
        limits: RunLimits,
    },
    ToolResult {
        run_id: &'a str,
        call_id: &'a str,
        #[serde(skip_serializing_if = "Option::is_none")]
        result: Option<Value>,
        #[serde(skip_serializing_if = "Option::is_none")]
        error: Option<String>,
    },
    Cancel {
        run_id: &'a str,
    },
    Shutdown,
}
#[derive(Debug, Serialize)]
pub struct HistoryTurn {
    pub role: &'static str,
    pub text: String,
}
#[derive(Debug, Clone, Copy, Serialize)]
pub struct RunLimits {
    pub turns: u32,
    pub output_tokens: u32,
    pub max_output_chars: u32,
    pub timeout_ms: u64,
    pub tool_timeout_ms: u64,
}

#[derive(Debug, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Outbound {
    Ready {
        protocol: u64,
        #[serde(default)]
        sdk: String,
    },
    TextDelta {
        run_id: String,
        text: String,
    },
    ToolCall {
        run_id: String,
        call_id: String,
        name: String,
        #[serde(default)]
        input: Value,
    },
    Completed {
        run_id: String,
        answer: String,
        #[serde(default)]
        stop_reason: String,
        #[serde(default)]
        model_calls: u32,
    },
    Failed {
        run_id: String,
        #[serde(default)]
        code: String,
        #[serde(default)]
        error: String,
    },
    Cancelled {
        run_id: String,
    },
}

pub struct Worker {
    child: Child,
    stdin: ChildStdin,
    messages: mpsc::Receiver<Outbound>,
    reader: JoinHandle<()>,
    logger: JoinHandle<()>,
    /// Run the worker is currently executing, until its terminal message arrives.
    active_run: Option<String>,
    pub sdk_version: String,
}

impl Worker {
    /// Start the worker and wait for its ready message. Model credentials are
    /// passed only through the child's private environment.
    pub async fn spawn(config: &Config) -> Result<Worker> {
        if !config.worker_script.is_file() {
            return Err(Error::Worker(format!(
                "Worker script {} is missing; run `npm ci && npm run build` in the worker folder",
                config.worker_script.display()
            )));
        }
        let mut command = Command::new(&config.node_bin);
        command
            .arg(&config.worker_script)
            .env_clear()
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .kill_on_drop(true);
        for key in [
            "PATH",
            "HOME",
            "LANG",
            "LC_ALL",
            "TMPDIR",
            "SSL_CERT_FILE",
            "NODE_EXTRA_CA_CERTS",
        ] {
            if let Ok(value) = std::env::var(key) {
                command.env(key, value);
            }
        }
        command.env("NODE_NO_WARNINGS", "1");
        if let Some(url) = &config.model_base_url {
            command.env("MODEL_BASE_URL", url);
        }
        if let Some(model) = &config.model_name {
            command.env("MODEL_NAME", model);
        }
        command.env("MODEL_API_KEY", &config.model_api_key);
        command.env(
            "MODEL_REQUEST_OPTIONS",
            config.model_request_options.to_string(),
        );
        command.env(
            "MODEL_TIMEOUT_SECONDS",
            config.model_timeout.as_secs().to_string(),
        );
        let mut child = command
            .spawn()
            .map_err(|e| Error::Worker(format!("Cannot start Node worker: {e}")))?;
        let stdin = child.stdin.take().expect("piped stdin");
        let stdout = child.stdout.take().expect("piped stdout");
        let stderr = child.stderr.take().expect("piped stderr");
        let (sender, messages) = mpsc::channel(256);
        let reader = tokio::spawn(async move {
            let mut lines = BufReader::new(stdout);
            let mut line = String::new();
            loop {
                line.clear();
                match lines.read_line(&mut line).await {
                    Ok(0) | Err(_) => break,
                    Ok(_) => {}
                }
                if line.len() > MAX_LINE_BYTES {
                    tracing::error!("Worker sent an oversized message; closing the channel");
                    break;
                }
                let trimmed = line.trim();
                if trimmed.is_empty() {
                    continue;
                }
                match serde_json::from_str::<Outbound>(trimmed) {
                    Ok(message) => {
                        if sender.send(message).await.is_err() {
                            break;
                        }
                    }
                    Err(_) => tracing::warn!("Worker sent an unrecognized message"),
                }
            }
        });
        let logger = tokio::spawn(async move {
            let mut lines = BufReader::new(stderr).lines();
            while let Ok(Some(line)) = lines.next_line().await {
                let line: String = line.chars().take(4000).collect();
                if !line.trim().is_empty() {
                    tracing::warn!(target: "backhaus_ai_backend::worker_stderr", "{line}");
                }
            }
        });
        let mut worker = Worker {
            child,
            stdin,
            messages,
            reader,
            logger,
            active_run: None,
            sdk_version: String::new(),
        };
        match tokio::time::timeout(READY_TIMEOUT, worker.next()).await {
            Ok(Some(Outbound::Ready { protocol, sdk })) if protocol == PROTOCOL_VERSION => {
                worker.sdk_version = sdk;
                Ok(worker)
            }
            Ok(Some(Outbound::Ready { protocol, .. })) => {
                worker.kill().await;
                Err(Error::Worker(format!(
                    "Worker protocol {protocol} does not match {PROTOCOL_VERSION}"
                )))
            }
            Ok(_) => {
                worker.kill().await;
                Err(Error::Worker(
                    "Worker exited before reporting ready; check its build and Node version".into(),
                ))
            }
            Err(_) => {
                worker.kill().await;
                Err(Error::Worker("Worker did not report ready in time".into()))
            }
        }
    }

    pub async fn send(&mut self, message: &Inbound<'_>) -> Result<()> {
        if let Inbound::Run { run_id, .. } = message {
            self.active_run = Some((*run_id).to_owned());
        }
        let mut line = serde_json::to_vec(message).map_err(|_| Error::Worker("Encode".into()))?;
        line.push(b'\n');
        self.stdin
            .write_all(&line)
            .await
            .map_err(|_| Error::Worker("Worker stdin closed".into()))?;
        self.stdin
            .flush()
            .await
            .map_err(|_| Error::Worker("Worker stdin closed".into()))
    }

    /// Next message, or `None` once the worker has exited or broken the protocol.
    pub async fn next(&mut self) -> Option<Outbound> {
        let message = self.messages.recv().await;
        if let Some(
            Outbound::Completed { run_id, .. }
            | Outbound::Failed { run_id, .. }
            | Outbound::Cancelled { run_id },
        ) = &message
            && self.active_run.as_deref() == Some(run_id.as_str())
        {
            self.active_run = None;
        }
        message
    }

    pub fn pid(&self) -> Option<u32> {
        self.child.id()
    }

    pub fn is_alive(&mut self) -> bool {
        matches!(self.child.try_wait(), Ok(None)) && !self.reader.is_finished()
    }

    /// Cancel an in-flight run and wait for its terminal message so that no
    /// stale output can be mistaken for a later run. Failure means the caller
    /// must replace this worker. A run that already finished needs nothing.
    pub async fn abort_run(&mut self, run_id: &str) -> Result<()> {
        if self.active_run.as_deref() != Some(run_id) {
            return Ok(());
        }
        self.send(&Inbound::Cancel { run_id }).await?;
        let drained = tokio::time::timeout(Duration::from_secs(5), async {
            while let Some(message) = self.next().await {
                match message {
                    Outbound::Completed { run_id: id, .. }
                    | Outbound::Failed { run_id: id, .. }
                    | Outbound::Cancelled { run_id: id } => {
                        if id == run_id {
                            return true;
                        }
                    }
                    Outbound::ToolCall {
                        run_id: id,
                        call_id,
                        ..
                    } if id == run_id => {
                        // Never execute tools for an abandoned run.
                        let _ = self
                            .send(&Inbound::ToolResult {
                                run_id,
                                call_id: &call_id,
                                result: None,
                                error: Some("Run cancelled".into()),
                            })
                            .await;
                    }
                    _ => {}
                }
            }
            false
        })
        .await;
        match drained {
            Ok(true) => Ok(()),
            _ => Err(Error::Worker(
                "Worker did not acknowledge cancellation".into(),
            )),
        }
    }

    pub async fn shutdown(mut self) {
        let _ = self.send(&Inbound::Shutdown).await;
        let _ = self.stdin.shutdown().await;
        if tokio::time::timeout(SHUTDOWN_GRACE, self.child.wait())
            .await
            .is_err()
        {
            tracing::warn!("Worker did not exit after shutdown; killing it");
            let _ = self.child.kill().await;
        }
        self.reader.abort();
        self.logger.abort();
    }

    async fn kill(&mut self) {
        let _ = self.child.kill().await;
        self.reader.abort();
        self.logger.abort();
    }
}

impl Drop for Worker {
    fn drop(&mut self) {
        // kill_on_drop covers the child; stop the pump tasks too.
        self.reader.abort();
        self.logger.abort();
    }
}
